File: src/NodeParser/AsExpressionNodeParser.ts
import { NodeParser } from "../NodeParser.js";
import ts from "typescript";
import { Context } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { BaseType } from "../Type/BaseType.js";

export class AsExpressionNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}

    public supportsNode(node: ts.AsExpression): boolean {
        return node.kind === ts.SyntaxKind.AsExpression;
    }
    public createType(node: ts.AsExpression, context: Context): BaseType | undefined {
        // only implement `as const` for now where we just ignore the as expression
        return this.childNodeParser.createType(node.expression, context);
    }
}

File: internal/parser/as_expression_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// AsExpressionNodeParser handles `expr as T` by parsing the expression and
// ignoring the asserted type; only `as const` is really supported
// (src/NodeParser/AsExpressionNodeParser.ts).
type AsExpressionNodeParser struct {
	childNodeParser NodeParser
}

func NewAsExpressionNodeParser(childNodeParser NodeParser) *AsExpressionNodeParser {
	return &AsExpressionNodeParser{childNodeParser: childNodeParser}
}

func (p *AsExpressionNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindAsExpression
}

func (p *AsExpressionNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	return p.childNodeParser.CreateType(node.AsAsExpression().Expression, ctx, nil)
}
